### ISSUE TITLE
Zato enmasse throws 'Boolean' object is not iterable' type error when using default values

### ISSUE THREAD

**Issue by jsabater** [zatosource/zato#891]
After adding the [default](https://github.com/zatosource/zato/commit/54a5ae9b3a41450558f5703401ec4656c36da05d) [value](https://github.com/zatosource/zato/commit/5b52924cadc9b057e5daa174def5ea195594e309) per attribute feature (issue #847), using **zato enmasse** produces errors.

Executing:

```
zato@kasumi:~$ zato enmasse /opt/zato/env/qs-1/server1/ --export-odb --verbose
could not fetch service list
Type channel_amqp has no 'get-list' service
Type web_socket has no 'get-list' service
Type pubsub_endpoint has no 'get-list' service
Type channel_jms_wmq has no 'get-list' service
Type channel_zmq has no 'get-list' service
Type def_sec has no 'get-list' service
Type http_soap has no 'get-list' service
Type scheduler has no 'get-list' service
Type notif_sql has no 'get-list' service
Type outconn_amqp has no 'get-list' service
Type outconn_jms_wmq has no 'get-list' service
Type query_cassandra has no 'get-list' service
ODB objects read
ODB objects merged in
Data exported to /opt/zato/zato-export-2018-10-11T08_16_05_586515.yml
```
Produces these in the logs:
```
2018-10-11 08:13:58,182 DEBG 'zato-server1' stdout output:
2018-10-11 08:13:58,181 - WARNING - 117:DummyThread-47 - zato.server.service:501 - Traceback (most recent call last):
  File "/opt/zato/3.0/code/zato-server/src/zato/server/service/__init__.py", line 476, in update_handle
    self._invoke(service, channel)
  File "/opt/zato/3.0/code/zato-server/src/zato/server/service/__init__.py", line 398, in _invoke
    service.handle()
  File "/opt/zato/3.0/code/zato-server/src/zato/server/service/internal/apispec/__init__.py", line 68, in handle
    include, exclude, self.request.input.query).get_info()
  File "/opt/zato/3.0/code/zato-server/src/zato/server/apispec/__init__.py", line 261, in get_info
    self.parse()
  File "/opt/zato/3.0/code/zato-server/src/zato/server/apispec/__init__.py", line 342, in parse
    info = ServiceInfo(details.name, details.service_class, self.simple_io_config)
  File "/opt/zato/3.0/code/zato-server/src/zato/server/apispec/__init__.py", line 90, in __init__
    self.parse()
  File "/opt/zato/3.0/code/zato-server/src/zato/server/apispec/__init__.py", line 95, in parse
    self.set_config()
  File "/opt/zato/3.0/code/zato-server/src/zato/server/apispec/__init__.py", line 182, in set_config
    self._add_ns_sio()
  File "/opt/zato/3.0/code/zato-server/src/zato/server/apispec/__init__.py", line 150, in _add_ns_sio
    for param in param_list:
TypeError: 'Boolean' object is not iterable


2018-10-11 08:13:58,182 DEBG 'zato-server1' stdout output:
2018-10-11 08:13:58,182 - WARNING - 117:DummyThread-47 - zato.server.service:585 - Could not invoke `zato.apispec.get-api-spec`, e:`Traceback (most recent call last):
  File "/opt/zato/3.0/code/zato-server/src/zato/server/service/__init__.py", line 583, in invoke_by_impl_name
    return self.update_handle(*invoke_args, **kwargs)
  File "/opt/zato/3.0/code/zato-server/src/zato/server/service/__init__.py", line 476, in update_handle
    self._invoke(service, channel)
  File "/opt/zato/3.0/code/zato-server/src/zato/server/service/__init__.py", line 398, in _invoke
    service.handle()
  File "/opt/zato/3.0/code/zato-server/src/zato/server/service/internal/apispec/__init__.py", line 68, in handle
    include, exclude, self.request.input.query).get_info()
  File "/opt/zato/3.0/code/zato-server/src/zato/server/apispec/__init__.py", line 261, in get_info
    self.parse()
  File "/opt/zato/3.0/code/zato-server/src/zato/server/apispec/__init__.py", line 342, in parse
    info = ServiceInfo(details.name, details.service_class, self.simple_io_config)
  File "/opt/zato/3.0/code/zato-server/src/zato/server/apispec/__init__.py", line 90, in __init__
    self.parse()
  File "/opt/zato/3.0/code/zato-server/src/zato/server/apispec/__init__.py", line 95, in parse
    self.set_config()
  File "/opt/zato/3.0/code/zato-server/src/zato/server/apispec/__init__.py", line 182, in set_config
    self._add_ns_sio()
  File "/opt/zato/3.0/code/zato-server/src/zato/server/apispec/__init__.py", line 150, in _add_ns_sio
    for param in param_list:
TypeError: 'Boolean' object is not iterable
`

2018-10-11 08:13:58,182 DEBG 'zato-server1' stdout output:
2018-10-11 08:13:58,182 - WARNING - 117:DummyThread-47 - zato.server.service:501 - Traceback (most recent call last):
  File "/opt/zato/3.0/code/zato-server/src/zato/server/service/__init__.py", line 476, in update_handle
    self._invoke(service, channel)
  File "/opt/zato/3.0/code/zato-server/src/zato/server/service/__init__.py", line 398, in _invoke
    service.handle()
  File "/opt/zato/3.0/code/zato-server/src/zato/server/service/internal/service.py", line 311, in handle
    response = func(id_, payload, channel, data_format, transport, serialize=True)
  File "/opt/zato/3.0/code/zato-server/src/zato/server/service/__init__.py", line 601, in invoke
    return self.invoke_by_impl_name(self.server.service_store.name_to_impl_name[name], *args, **kwargs)
  File "/opt/zato/3.0/code/zato-server/src/zato/server/service/__init__.py", line 583, in invoke_by_impl_name
    return self.update_handle(*invoke_args, **kwargs)
  File "/opt/zato/3.0/code/zato-server/src/zato/server/service/__init__.py", line 476, in update_handle
    self._invoke(service, channel)
  File "/opt/zato/3.0/code/zato-server/src/zato/server/service/__init__.py", line 398, in _invoke
    service.handle()
  File "/opt/zato/3.0/code/zato-server/src/zato/server/service/internal/apispec/__init__.py", line 68, in handle
    include, exclude, self.request.input.query).get_info()
  File "/opt/zato/3.0/code/zato-server/src/zato/server/apispec/__init__.py", line 261, in get_info
    self.parse()
  File "/opt/zato/3.0/code/zato-server/src/zato/server/apispec/__init__.py", line 342, in parse
    info = ServiceInfo(details.name, details.service_class, self.simple_io_config)
  File "/opt/zato/3.0/code/zato-server/src/zato/server/apispec/__init__.py", line 90, in __init__
    self.parse()
  File "/opt/zato/3.0/code/zato-server/src/zato/server/apispec/__init__.py", line 95, in parse
    self.set_config()
  File "/opt/zato/3.0/code/zato-server/src/zato/server/apispec/__init__.py", line 182, in set_config
    self._add_ns_sio()
  File "/opt/zato/3.0/code/zato-server/src/zato/server/apispec/__init__.py", line 150, in _add_ns_sio
    for param in param_list:
TypeError: 'Boolean' object is not iterable


2018-10-11 08:13:58,183 DEBG 'zato-server1' stdout output:
2018-10-11 08:13:58,183 - ERROR - 117:DummyThread-47 - zato.server.connection.http_soap.channel:324 - Caught an exception, cid:`bf413d6ff5fd6b63111318fa`, status_code:`500`, _format_exc:`Traceback (most recent call last):
  File "/opt/zato/3.0/code/zato-server/src/zato/server/connection/http_soap/channel.py", line 268, in dispatch
    payload, worker_store, self.simple_io_config, post_data, path_info, soap_action)
  File "/opt/zato/3.0/code/zato-server/src/zato/server/connection/http_soap/channel.py", line 502, in handle
    params_priority=channel_item.params_pri)
  File "/opt/zato/3.0/code/zato-server/src/zato/server/service/__init__.py", line 476, in update_handle
    self._invoke(service, channel)
  File "/opt/zato/3.0/code/zato-server/src/zato/server/service/__init__.py", line 398, in _invoke
    service.handle()
  File "/opt/zato/3.0/code/zato-server/src/zato/server/service/internal/service.py", line 311, in handle
    response = func(id_, payload, channel, data_format, transport, serialize=True)
  File "/opt/zato/3.0/code/zato-server/src/zato/server/service/__init__.py", line 601, in invoke
    return self.invoke_by_impl_name(self.server.service_store.name_to_impl_name[name], *args, **kwargs)
  File "/opt/zato/3.0/code/zato-server/src/zato/server/service/__init__.py", line 583, in invoke_by_impl_name
    return self.update_handle(*invoke_args, **kwargs)
  File "/opt/zato/3.0/code/zato-server/src/zato/server/service/__init__.py", line 476, in update_handle
    self._invoke(service, channel)
  File "/opt/zato/3.0/code/zato-server/src/zato/server/service/__init__.py", line 398, in _invoke
    service.handle()
  File "/opt/zato/3.0/code/zato-server/src/zato/server/service/internal/apispec/__init__.py", line 68, in handle
    include, exclude, self.request.input.query).get_info()
  File "/opt/zato/3.0/code/zato-server/src/zato/server/apispec/__init__.py", line 261, in get_info
    self.parse()
  File "/opt/zato/3.0/code/zato-server/src/zato/server/apispec/__init__.py", line 342, in parse
    info = ServiceInfo(details.name, details.service_class, self.simple_io_config)
  File "/opt/zato/3.0/code/zato-server/src/zato/server/apispec/__init__.py", line 90, in __init__
    self.parse()
  File "/opt/zato/3.0/code/zato-server/src/zato/server/apispec/__init__.py", line 95, in parse
    self.set_config()
  File "/opt/zato/3.0/code/zato-server/src/zato/server/apispec/__init__.py", line 182, in set_config
    self._add_ns_sio()
  File "/opt/zato/3.0/code/zato-server/src/zato/server/apispec/__init__.py", line 150, in _add_ns_sio
    for param in param_list:
TypeError: 'Boolean' object is not iterable
```

Hot-deploying the affected module **login.py** does not produce any errors and the list service can be called without problems using **curl** from the console.

Here is the current version of the schema:

```
# coding: utf8
from __future__ import absolute_import, division
from __future__ import print_function, unicode_literals
from .base import Base
from sqlalchemy import Column, Boolean, Integer, String, Index


class Login(Base):
    __tablename__ = 'login'
    __rels__ = []
    __table_args__ = (
        # Trigram GIN indexes for searches (using ILIKE)
        Index('ix_trgm_login_username', 'username', postgresql_using='gin',
              postgresql_ops={'username': 'gin_trgm_ops'}),
        Index('ix_trgm_login_name', 'name', postgresql_using='gin',
              postgresql_ops={'name': 'gin_trgm_ops'}),
        Index('ix_trgm_login_surname', 'surname', postgresql_using='gin',
              postgresql_ops={'surname': 'gin_trgm_ops'}),
    )

    id = Column(Integer, primary_key=True)
    username = Column(String(20), index=True, unique=True, nullable=False)
    password = Column(String(255), nullable=False)
    name = Column(String(50), index=True)
    surname = Column(String(50), index=True)
    email = Column(String(255), index=True, unique=True)
    is_admin = Column(Boolean, default=False)

    def __repr__(self):
        return "<Login(id='%s', username='%s', name='%s', surname='%s', email='%s')>" % (
            self.id, self.username, self.name, self.surname, self.email)
```

And here is the content of the [login module](https://bitbucket.org/jsabater/genesisng/src/default/genesisng/services/login.py), which includes several services:

```
# -*- coding: utf-8 -*-
from __future__ import absolute_import, division
from __future__ import print_function, unicode_literals
from contextlib import closing
from httplib import OK, NO_CONTENT, CREATED, NOT_FOUND, CONFLICT
from zato.server.service import Service, Boolean, Integer, AsIs
from genesisng.schema.login import Login
from sqlalchemy import or_, and_, func
from sqlalchemy.exc import IntegrityError
from urlparse import parse_qs


class Get(Service):
    """Service class to get a login by id."""
    """Channel /genesisng/logins/{id}/get."""

    class SimpleIO:
        input_required = (Integer('id'),)
        output_required = ('id', 'username')
        output_optional = ('password', 'name', 'surname', 'email', 'is_admin')
        skip_empty_keys = True

    def handle(self):
        conn = self.user_config.genesisng.database.connection
        id_ = self.request.input.id

        with closing(self.outgoing.sql.get(conn).session()) as session:
            result = session.query(Login).filter(Login.id == id_).one_or_none()

            if result:
                self.response.status_code = OK
                self.response.payload = result
            else:
                self.response.status_code = NOT_FOUND
                self.response.payload = ''


class Validate(Service):
    """Service class to validate credentials."""
    """Channel /genesisng/logins/validate."""

    class SimpleIO:
        input_required = ('username', AsIs('password'))
        output_required = ('id', 'username')
        output_optional = ('password', 'name', 'surname', 'email', 'is_admin')
        skip_empty_keys = True

    def handle(self):
        conn = self.user_config.genesisng.database.connection
        username = self.request.input.username
        password = self.request.input.password

        with closing(self.outgoing.sql.get(conn).session()) as session:
            result = session.query(Login).\
                filter(and_(Login.username == username,
                            Login.password == password)).\
                one_or_none()

            if result:
                self.response.status_code = OK
                self.response.payload = result
            else:
                self.response.status_code = NOT_FOUND
                self.response.payload = ''


class Create(Service):
    """Service class to create a new login."""
    """Channel /genesisng/logins/create."""

    class SimpleIO:
        input_required = ('username', AsIs('password'))
        input_optional = ('name', 'surname', 'email',
                          Boolean('is_admin', default=False))
        output_required = ('id', 'username')
        output_optional = ('password', 'name', 'surname', 'email', 'is_admin')
        skip_empty_keys = True

    def handle(self):
        # TODO: Use Cerberus to validate input?
        # http://docs.python-cerberus.org/en/stable/
        conn = self.user_config.genesisng.database.connection

        p = self.request.input
        login = Login(
            username=p.username,
            password=p.password,
            name=p.name,
            surname=p.surname,
            email=p.email,
            is_admin=p.is_admin)

        with closing(self.outgoing.sql.get(conn).session()) as session:
            try:
                session.add(login)
                session.commit()
                self.response.status_code = CREATED
                self.response.payload = login
                url = self.user_config.genesisng.location.logins
                self.response.headers['Location'] = url.format(id=login.id)

            except IntegrityError:
                # Constraint prevents duplication of username or emails.
                session.rollback()
                self.response.status_code = CONFLICT
                # TODO: Return well-formed error response
                # https://medium.com/@suhas_chatekar/return-well-formed-error-responses-from-your-rest-apis-956b5275948
                self.response.payload = ''


class Delete(Service):
    """Service class to delete an existing login."""
    """Channel /genesisng/logins/{id}/delete."""

    class SimpleIO:
        input_required = (Integer('id'),)

    def handle(self):
        conn = self.user_config.genesisng.database.connection
        id_ = self.request.input.id

        with closing(self.outgoing.sql.get(conn).session()) as session:
            deleted = session.query(Login).filter(Login.id == id_).delete()
            session.commit()

            if deleted:
                self.response.status_code = NO_CONTENT
                self.response.payload = ''
            else:
                self.response.status_code = NOT_FOUND
                self.response.payload = ''


class Update(Service):
    """Service class to update an existing login."""
    """Channel /genesisng/logins/{id}/update."""

    class SimpleIO:
        input_required = (Integer('id'),)
        input_optional = ('username', AsIs('password'), 'name', 'surname',
                          'email', Boolean('is_admin'))
        output_required = ('id', 'username')
        output_optional = ('password', 'name', 'surname', 'email', 'is_admin')
        skip_empty_keys = True

    def handle(self):
        conn = self.user_config.genesisng.database.connection
        id_ = self.request.input.id
        p = self.request.input

        with closing(self.outgoing.sql.get(conn).session()) as session:
            result = session.query(Login).filter(Login.id == id_).one_or_none()

            if result:
                # Update dictionary keys
                if p.username:
                    result.username = p.username
                if p.password:
                    result.password = p.password
                if p.name:
                    result.name = p.name
                if p.surname:
                    result.surname = p.surname
                if p.email:
                    result.email = p.email
                if p.is_admin != '':
                    result.is_admin = p.is_admin
                session.commit()
                self.response.status_code = OK
                self.response.payload = result
            else:
                self.response.status_code = NOT_FOUND
                self.response.payload = ''


class List(Service):
    """Service class to get a list of all logins in the system."""
    """Channel /genesisng/logins/list."""

    class SimpleIO:
        input_optional = (Integer('page'), Integer('size'), 'sort_by',
                          'order_by', 'filters', 'search', 'fields')
        output_required = ('count', )
        output_optional = ('id', 'username', 'password', 'name', 'surname',
                           'email', 'is_admin')
        output_repeated = True
        skip_empty_keys = True

    def handle(self):
        conn = self.user_config.genesisng.database.connection
        default_page_size = int(
            self.user_config.genesisng.database.default_page_size)
        max_page_size = int(self.user_config.genesisng.database.max_page_size)
        # TODO: Have a default order_by and sort_by in the KVDB?
        default_order_by = 'id'
        default_sort_by = 'asc'

        # Pagination is always mandatory
        page = 1
        size = default_page_size

        # Sorting is always mandatory
        order_by = default_order_by
        sort_by = default_sort_by

        # Filtering is optional
        # Format: filters=field|operator|value (multiple)
        filters = []

        # Fields projection is optional
        # Format: fields=field (multiple)
        fields = []

        # Search is optional
        search = None

        # Check for parameters in the query string
        qs = parse_qs(self.wsgi_environ['QUERY_STRING'])
        if qs:

            # Handle pagination
            try:
                page = int(qs['page'][0])
                page = 1 if page < 1 else page

                size = int(qs['size'][0])
                size = default_page_size if size < 1 else size
                size = default_page_size if size > max_page_size else size
            except (ValueError, KeyError, IndexError):
                # Assume default values instead of returning 400 Bad Request
                pass

            # Handle sorting
            try:
                order_by = qs['order_by'][0].lower()
                # Fields allowed for ordering are id, username, name, surname
                # and email
                if order_by not in ('id', 'username', 'name', 'surname',
                                    'email'):
                    order_by = default_order_by

                sort_by = qs['sort_by'][0].lower()
                sort_by = default_sort_by if sort_by not in (
                    'asc', 'desc') else sort_by
            except (ValueError, KeyError, IndexError):
                # Assume default values instead of returning 400 Bad Request
                pass

            # Handle filtering
            try:
                for f in qs['filters']:
                    field, operator, value = f.split('|')
                    if field not in ('id', 'username', 'password', 'name',
                                     'surname', 'email', 'is_admin'):
                        raise ValueError(
                            'Field %s is not allowed for filtering' % field)
                    if operator not in ('lt', 'lte', 'eq', 'ne', 'gte', 'gt'):
                        raise ValueError(
                            'Operator %s is not allowed for filtering' %
                            operator)
                    filters.append((field, operator, value))
            except (ValueError, KeyError):
                # Do not apply any filtering instead of returning 400 Bad
                # Request
                pass

            # Handle search
            try:
                search = qs['search'][0]
            except (ValueError, KeyError):
                # Do not search for terms instead of returning 400 Bad Request
                pass

            # Handle fields projection
            try:
                for field in qs['fields']:
                    if field not in ('id', 'username', 'password', 'name',
                                     'surname', 'email', 'is_admin'):
                        raise ValueError(
                            'Field %s is not allowed for projection' % field)
                    fields.append(field)
            except (ValueError, KeyError):
                # Do not apply any fields projection instead of returning 400
                # Bad Request
                pass

        # Calculate limit and offset
        limit = size
        offset = size * (page - 1)

        # Calculate criteria and direction
        criteria = order_by
        direction = sort_by

        # Prepare filters
        # TODO: Use sqlalchemy-filters?
        # https://pypi.org/project/sqlalchemy-filters/
        conditions = []
        for f in filters:
            field, operator, value = f
            if operator == 'lt':
                conditions.append(Login.__table__.columns[field] < value)
            elif operator == 'lte':
                conditions.append(Login.__table__.columns[field] <= value)
            elif operator == 'eq':
                conditions.append(Login.__table__.columns[field] == value)
            elif operator == 'ne':
                conditions.append(Login.__table__.columns[field] != value)
            elif operator == 'gte':
                conditions.append(Login.__table__.columns[field] >= value)
            elif operator == 'gt':
                conditions.append(Login.__table__.columns[field] > value)

        # Prepare search
        if search:
            term = '%' + search + '%'

        # Prepare fields projection
        columns = []
        if not fields:
            fields = ('id', 'username', 'password', 'name', 'surname', 'email',
                      'is_admin')
        columns = [Login.__table__.columns[f] for f in fields]

        # Execute query
        with closing(self.outgoing.sql.get(conn).session()) as session:
            query = session.query(func.count().over().label('count'))
            for c in columns:
                query = query.add_columns(c)
            for c in conditions:
                query = query.filter(c)
            if search:
                query = query.filter(
                    or_(
                        Login.username.ilike(term), Login.name.ilike(term),
                        Login.surname.ilike(term), Login.email.ilike(term)))
            if direction == 'asc':
                query = query.order_by(Login.__table__.columns[criteria].asc())
            else:
                query = query.order_by(
                    Login.__table__.columns[criteria].desc())
            query = query.offset(offset)
            query = query.limit(limit)
            result = query.all()
            self.response.payload[:] = result if result else []
```

**Comment by dsuch** [comment]
I cannot reproduce it with the code sent in. The traceback suggests that there was a service with a SimpleIO definition akin to:

```python
class SimpleIO:
  input_required = Boolean('elem_name')
```

This could be input_optional possibly, I am not sure from the exception alone. But the key point is that there was a missing comma in the definition.

I can work on not requiring that extra comma but I do not see a connection between this error and using default values.

I will close the ticket once I implement the change that will drop the requirement to add commas, feel free to re-open if you can prepare a small test case displaying the behaviour originally reported (default values causing the error).

**Comment by dsuch** [comment]
Done, the extra comma is no longer needed.

**Comment by jsabater** [comment]
My fault. I only had channels for the services in the `login.py` module, but other services had been previously deployed although remained _unused_. I didn't have in mind that **zato enmasse** would be taking them into account.

Using grep I was finally able to find the module with the definition without the required comma (`rate.py`). I fixed it and **zato enmasse** produced the desired output again.

Regarding the patch, I updated the code, removed the trailing commas in all the services of the `login.py`module, hot-deployed and tested it and it's working fine.

Thanks!
